### ISSUE TITLE
PerforatedAI 

### ISSUE THREAD

**Issue by RorryB** [pytorch-fdn/ecosystem#54]
### Contact emails

rorry@perforatedai.com,Mark.Westerlund.Dev@gmail.com

### Project summary

Dendritic Optimization, for smarter, smaller, cheaper ML Models

### Project description

Dendritic Optimization empowers ML engineers to build smarter, smaller, and more accurate neural networks. 

Today’s neural networks are informed by century-old models of biological intelligence. Neuroscientists now recognize that biological dendrites, not neurons, are the core building blocks of the way our brains function. 

This repo introduces artificial dendrites into existing neural-networks, unlocking compression levels of up to 90% while maintaining or even improving accuracy. Artificial dendrites emulate the non-linear integration and processing capabilities of biological dendrites, allowing artificial neuron modules to perform complex computations. They eliminate the most extreme cost/accuracy tradeoffs while building a model before pushing to the edge or deploying to the cloud. 

This breakthrough offers a new path forward for ML engineers and data scientists, overcoming the economic and technical limitations of legacy neural networks. With just minutes of coding and no changes to your core architecture, dendrite-enhanced neurons can be integrated into exiting training pipelines using this open source repo.

### Are there any other projects in the PyTorch Ecosystem similar to yours?  If, yes, what are they?

No.

### Project repo URL

https://github.com/PerforatedAI/PerforatedAI

### Project license

Apache License 2.0

### GitHub handles of the project maintainer(s)

@RorryB,@MwesterlundDev

### Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

Perforated AI

### Website URL

https://www.perforatedai.com/

### Documentation

https://github.com/PerforatedAI/PerforatedAI/tree/main/API

### How do you build and test the project today (continuous integration)?  Please describe.

Build is done automatically when new tags are checked in.  Testing is currently done manually.  Would like to work with the team as we build our first automatic tests to ensure we meet the requirements.

### Version of PyTorch

Works with all PyTorch versions > 1.12

### Components of PyTorch

Not 100% sure how to answer this question for our repo, but here is what Claude suggested:

What components of PyTorch does your project leverage?
PerforatedAI deeply integrates with PyTorch's core architecture through the following components:

1. Module System
nn.Module: Base class for all custom components (PAINeuronModule, TrackedNeuronModule, PAIDendriteModule, DendriteValueTracker, PAISequential, ResNetPAI)
2. Parameters & Buffers
nn.Parameter: Learnable dendrite weights
nn.ParameterList: Collections of dendrite-to-neuron and dendrite-to-dendrite connection weights
register_buffer(): Non-trainable state storage (18+ uses) for shape tracking, dimension info, cycle counts, and statistics
3. Autograd Integration
register_hook(): Gradient interception during backward pass
Enables per-dendrite gradient tracking and manipulation
4. Supported Layer Types
Configured to wrap: nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d
Flexibility to wrap any module with additional settings
5. Tensor Operations
Dtype management: Dynamic dtype conversion based on GPA.pc.d_type
6. Serialization
state_dict() / load_state_dict(): Core model persistence (20+ uses)
torch.save(): Fallback serialization 
Safetensors integration: Efficient alternative to pickle-based saving
7. Training Infrastructure
Optimizer API: Integrates with any PyTorch optimizer
Scheduler API: Learning rate schedule management
Device management: CUDA support and multi-GPU handling
8. Initialization Methods
torch.nn.init: Weight initialization strategies (kaiming_uniform_, xavier_uniform_, etc.)
9. Functional API
torch.nn.functional: Used for operations like F.linear, F.conv2d

Architecture Pattern
The project uses a wrapper-based architecture that intercepts PyTorch's forward and backward passes to add dendrite functionality to any compatible PyTorch module, while maintaining full compatibility with standard PyTorch training loops and tools.

### How long do you expect to maintain the project?

Indefinitely.

### Additional information

We have recently hit the Measurable Requirements and added the Github Layout Requirements.  I do not believe we have all of the Functional Requirements yet. Would like to start the discussion with the team to make sure that is done correctly from the beginning in the right way.

**Comment by jspisak** [comment]
Thanks for submitting this @RorryB - can you take a look at the areas I call out below and let me know if we can fix / close them?


Requirement | Status
-- | --
PyTorch Support | ✅
CI Workflow | ✅
Governance | ✅
User Experience | ✅
Quick Start | ✅
Documentation | 👍 Present
Permissive License | ✅
Functional Testing | ❓ Unknown
Packages | ❌ No published releases
Ongoing Maintenance | 👍 Active
Core Maintainers | ✅
Active Contributors | ✅
90d Commits | ✅
≥200 Stars | ✅
LICENSE | ✅
README | ✅
CONTRIBUTING.md | ✅
CODEOWNERS.md | ✅
CODE_OF_CONDUCT.md | ✅

**Comment by RorryB** [comment]
Hey @jspisak , thanks for the review!  I believe packages can be closed: [link](https://pypi.org/project/perforatedai/).

Functional testing is correct that we do not have that yet.  Is there any documentation with more specifics about the types of tests we should have?  Or is this flexible and we can decide on our side what tests are best to ensure continued functionality with updated versions of PyTorch?  Happy to get that done as soon as possible once we know how to proceed.

**Comment by jspisak** [comment]
I think it's up to you all to decide as maintainers such that you all are delivering a high quality project to the community. 

On the packaging, are you planning to release UV or pypi packages?

**Comment by RorryB** [comment]
Great, we will get started on the functional tests and let you know when they're set up.

PyPi Package is here: https://pypi.org/project/perforatedai/.  Let me know if this it not correct.

**Comment by jspisak** [comment]
perfect - thanks @RorryB !

**Comment by michellearoth** [comment]
Approved! 

<img width="727" height="971" alt="Image" src="https://github.com/user-attachments/assets/3ef24636-9223-4f1e-a7a4-9cc9b3f0668e" />
